### ISSUE TITLE
permission denied, scandir '/lost+found'

### ISSUE THREAD

**Issue by mickaelistria** [sourcegraph/javascript-typescript-langserver#129]
I'm trying to connect Eclipse LSP4E with this Language Server.
I've run the build instructions without issues, and LSP4E It basically starts `node /home/mistria/git/javascript-typescript-langserver/lib/language-server-stdio.js` and connects to stdio streams.

The server says it's failing because of some `permission denied, scandir '/lost+found'`, see full conversation below.

```
!ENTRY org.eclipse.lsp4e 1 0 2017-04-04 19:01:19.852
!MESSAGE RequestMessage
{
  "jsonrpc": "2.0",
  "id": "1",
  "method": "initialize",
  "params": {
    "rootUri": "file:/home/mistria/git/mathjs/",
    "capabilities": {
      "workspace": {
        "applyEdit": true,
        "symbol": {},
        "executeCommand": {}
      },
      "textDocument": {
        "synchronization": {
          "willSave": true,
          "willSaveWaitUntil": true,
          "didSave": true
        },
        "completion": {
          "completionItem": {
            "snippetSupport": true
          }
        },
        "hover": {},
        "signatureHelp": {},
        "references": {},
        "documentHighlight": {},
        "documentSymbol": {},
        "formatting": {},
        "rangeFormatting": {},
        "definition": {},
        "codeAction": {},
        "codeLens": {},
        "documentLink": {},
        "rename": {}
      }
    },
    "clientName": "Eclipse Platform"
  }
}
glob error { Error: EACCES: permission denied, scandir '/lost+found'
  errno: -13,
  code: 'EACCES',
  syscall: 'scandir',
  path: '/lost+found' }

!ENTRY org.eclipse.lsp4e 1 0 2017-04-04 19:01:20.598
!MESSAGE ResponseMessage
{
  "jsonrpc": "2.0",
  "id": "1",
  "result": {
    "capabilities": {
      "textDocumentSync": 1,
      "hoverProvider": true,
      "completionProvider": {
        "resolveProvider": false,
        "triggerCharacters": [
          "."
        ]
      },
      "definitionProvider": true,
      "referencesProvider": true,
      "documentSymbolProvider": true,
      "workspaceSymbolProvider": true
    }
  }
}

!ENTRY org.eclipse.lsp4e 1 0 2017-04-04 19:01:20.608
!MESSAGE NotificationMessage
{
  "jsonrpc": "2.0",
  "method": "textDocument/didOpen",
  "params": {
    "textDocument": {
      "uri": "file:/home/mistria/git/mathjs/lib/constants.js",
      "languageId": "js",
      "version": 1,
      "text": "\u0027use strict\u0027;\n\nvar object \u003d require(\u0027./utils/object\u0027);\nvar bigConstants \u003d require(\u0027./utils/bignumber/constants\u0027);\n\nfunction factory (type, config, load, typed, math) {\n  // listen for changed in the configuration, automatically reload\n  // constants when needed\n  math.on(\u0027config\u0027, function (curr, prev) {\n    if (curr.number !\u003d\u003d prev.number) {\n      factory(type, config, load, typed, math);\n    }\n  });\n\n  math[\u0027true\u0027]     \u003d true;\n  math[\u0027false\u0027]    \u003d false;\n  math[\u0027null\u0027]     \u003d null;\n  math[\u0027uninitialized\u0027] \u003d require(\u0027./utils/array\u0027).UNINITIALIZED;\n\n  if (config.number \u003d\u003d\u003d \u0027bignumber\u0027) {\n    math[\u0027Infinity\u0027] \u003d new type.BigNumber(Infinity);\n    math[\u0027NaN\u0027]      \u003d new type.BigNumber(NaN);\n\n    object.lazy(math, \u0027pi\u0027,  function () {return bigConstants.pi(type.BigNumber)});\n    object.lazy(math, \u0027tau\u0027, function () {return bigConstants.tau(type.BigNumber)});\n    object.lazy(math, \u0027e\u0027,   function () {return bigConstants.e(type.BigNumber)});\n    object.lazy(math, \u0027phi\u0027, function () {return bigConstants.phi(type.BigNumber)}); // golden ratio, (1+sqrt(5))/2\n\n    // uppercase constants (for compatibility with built-in Math)\n    object.lazy(math, \u0027E\u0027,       function () {return math.e;});\n    object.lazy(math, \u0027LN2\u0027,     function () {return new type.BigNumber(2).ln();});\n    object.lazy(math, \u0027LN10\u0027,    function () {return new type.BigNumber(10).ln()});\n    object.lazy(math, \u0027LOG2E\u0027,   function () {return new type.BigNumber(1).div(new type.BigNumber(2).ln());});\n    object.lazy(math, \u0027LOG10E\u0027,  function () {return new type.BigNumber(1).div(new type.BigNumber(10).ln())});\n    object.lazy(math, \u0027PI\u0027,      function () {return math.pi});\n    object.lazy(math, \u0027SQRT1_2\u0027, function () {return new type.BigNumber(\u00270.5\u0027).sqrt()});\n    object.lazy(math, \u0027SQRT2\u0027,   function () {return new type.BigNumber(2).sqrt()});\n  }\n  else {\n    math[\u0027Infinity\u0027] \u003d Infinity;\n    math[\u0027NaN\u0027]      \u003d NaN;\n\n    math.pi  \u003d Math.PI;\n    math.tau \u003d Math.PI * 2;\n    math.e   \u003d Math.E;\n    math.phi \u003d 1.61803398874989484820458683436563811772030917980576286213545; // golden ratio, (1+sqrt(5))/2\n\n    // uppercase constants (for compatibility with built-in Math)\n    math.E           \u003d math.e;\n    math.LN2         \u003d Math.LN2;\n    math.LN10        \u003d Math.LN10;\n    math.LOG2E       \u003d Math.LOG2E;\n    math.LOG10E      \u003d Math.LOG10E;\n    math.PI          \u003d math.pi;\n    math.SQRT1_2     \u003d Math.SQRT1_2;\n    math.SQRT2       \u003d Math.SQRT2;\n    \n  }\n\n  // complex i\n  math.i \u003d new type.Complex(0, 1);\n\n  // meta information\n  math.version \u003d require(\u0027./version\u0027);\n}\n\nexports.factory \u003d factory;\nexports.lazy \u003d false;  // no lazy loading of constants, the constants themselves are lazy when needed\nexports.math \u003d true;   // request access to the math namespace"
    }
  }
}

!ENTRY org.eclipse.lsp4e 1 0 2017-04-04 19:01:20.703
!MESSAGE NotificationMessage
{
  "jsonrpc": "2.0",
  "method": "window/logMessage",
  "params": {
    "type": 1,
    "message": "Background fetching failed  { Error: EACCES: permission denied, scandir \u0027/lost+found\u0027\n  errno: -13,\n  code: \u0027EACCES\u0027,\n  syscall: \u0027scandir\u0027,\n  path: \u0027/lost+found\u0027 }"
  }
}
glob error { Error: EACCES: permission denied, scandir '/lost+found'
  errno: -13,
  code: 'EACCES',
  syscall: 'scandir',
  path: '/lost+found' }

!ENTRY org.eclipse.lsp4e 1 0 2017-04-04 19:01:20.731
!MESSAGE NotificationMessage
{
  "jsonrpc": "2.0",
  "method": "window/logMessage",
  "params": {
    "type": 1,
    "message": "Failed to fetch module structure:  { Error: EACCES: permission denied, scandir \u0027/lost+found\u0027\n  errno: -13,\n  code: \u0027EACCES\u0027,\n  syscall: \u0027scandir\u0027,\n  path: \u0027/lost+found\u0027 }"
  }
}
(node:25481) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: EACCES: permission denied, scandir '/lost+found'
(node:25481) DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

!ENTRY org.eclipse.lsp4e 1 0 2017-04-04 19:01:21.598
!MESSAGE RequestMessage
{
  "jsonrpc": "2.0",
  "id": "2",
  "method": "textDocument/completion",
  "params": {
    "textDocument": {
      "uri": "file:/home/mistria/git/mathjs/lib/constants.js"
    },
    "uri": "file:/home/mistria/git/mathjs/lib/constants.js",
    "position": {
      "line": 56,
      "character": 4
    }
  }
}
glob error { Error: EACCES: permission denied, scandir '/lost+found'
  errno: -13,
  code: 'EACCES',
  syscall: 'scandir',
  path: '/lost+found' }

!ENTRY org.eclipse.lsp4e 1 0 2017-04-04 19:01:21.601
!MESSAGE NotificationMessage
{
  "jsonrpc": "2.0",
  "method": "window/logMessage",
  "params": {
    "type": 1,
    "message": "Failed to fetch module structure:  { Error: EACCES: permission denied, scandir \u0027/lost+found\u0027\n  errno: -13,\n  code: \u0027EACCES\u0027,\n  syscall: \u0027scandir\u0027,\n  path: \u0027/lost+found\u0027 }"
  }
}

!ENTRY org.eclipse.lsp4e 4 0 2017-04-04 19:01:21.605
!MESSAGE Request textDocument/completion failed with message: EACCES: permission denied, scandir '/lost+found'
!STACK 0
org.eclipse.lsp4j.jsonrpc.ResponseErrorException: Request textDocument/completion failed with message: EACCES: permission denied, scandir '/lost+found'
```

**Comment by alexsaveliev** [comment]
Duplicate of #114

**Comment by felixfbecker** [comment]
```
"rootUri": "file:/home/mistria/git/mathjs/",
```
that doesn't look like a usual file URL

**Comment by mickaelistria** [comment]
`file;` URIs are what Java use as default and works for all software I've seen so far:
```
mistria@mistria-rh:~/git/javascript-typescript-langserver$ jjs
jjs> new java.io.File("/home/mistria").toURI().toString()
file:/home/mistria/
```
VSCode Language Servers (CSS, JSON, HTML) and OmniSharp-node-client do support this URL properly.
The `vscode-language-server` module provide a `Files.uriToFilePath(...)` operation that works with such URIs. Maybe you could use it instead of `LocalFileSystem.resolveUriToPath(...)`.

**Comment by mickaelistria** [comment]
I'm using 2.0.0-pre8 ( 0f35711f0b3ea7d4057fa4043df888dbdd1125ce )

**Comment by alexsaveliev** [comment]
@mickaelistria is there any instructions how do you connect Eclipse LSP4E with JSTS lang server? I'd like to take a closer look into it

- what Eclipse version I should use? (seems to be Oxigen but I'd like to make sure)
- If I understood it properly I should somehow add new launch configuration and associate it with the content type. Where and how should I do it? Maybe there is a sample somewhere?

Thanks!

**Comment by mickaelistria** [comment]
It requires to have Oxygen and to install LSP4E in it.
The code (work in progress) is in https://github.com/mickaelistria/eclipse-bluesky . But using it to test some changes in the server do require knowledge in RCP/plugin development. It's about replace the `" "` in https://github.com/mickaelistria/eclipse-bluesky/blob/master/org.eclipse.bluesky/src/org/eclipse/bluesky/JSLanguageServer.java#L11 by the full path to language-server-stdio.js.
There is also a way to do it via the Launch Configuration as you discovered. It consists in creating a Launch Configuration (external one typically, running `/usr/bin/node /path/to/language-server-stdio.js`) and then use the preference page to associate it with the js content-type, and then to open a .js file with the `Generic Editor`. It's not much easier...
But if you want me to test some patches, please ping me on the related PRs and I'll try to react promptly.

**Comment by mickaelistria** [comment]
@felixfbecker BTW, I believe the scheme `file://` is for URLs (RFC 1738), not for URIs. The LSP deals with URIs.

**Comment by felixfbecker** [comment]
Afaik URLs are a subset of URIs, so every valid URL is a valid URI. Not saying the // is required, typically it indicates a host except when a third slash follows, just saying it's uncommon

**Comment by mickaelistria** [comment]
Actually, I also tried with a `file:///home/mistria/git/mathjs` uri and I got some similar error

```
!ENTRY org.eclipse.lsp4e 1 0 2017-04-05 08:37:48.212
!MESSAGE RequestMessage
{
  "jsonrpc": "2.0",
  "id": "1",
  "method": "initialize",
  "params": {
    "rootUri": "file:///home/mistria/git/mathjs",
    "capabilities": {
      "workspace": {
        "applyEdit": true,
        "symbol": {},
        "executeCommand": {}
      },
      "textDocument": {
        "synchronization": {
          "willSave": true,
          "willSaveWaitUntil": true,
          "didSave": true
        },
        "completion": {
          "completionItem": {
            "snippetSupport": true
          }
        },
        "hover": {},
        "signatureHelp": {},
        "references": {},
        "documentHighlight": {},
        "documentSymbol": {},
        "formatting": {},
        "rangeFormatting": {},
        "definition": {},
        "codeAction": {},
        "codeLens": {},
        "documentLink": {},
        "rename": {}
      }
    },
    "clientName": "Eclipse Platform"
  }
}

!ENTRY
glob error { Error: EACCES: permission denied, scandir '/lost+found'
  errno: -13,
  code: 'EACCES',
  syscall: 'scandir',
  path: '/lost+found' }
``` 

Are there some options to try on the launcher?

**Comment by alexsaveliev** [comment]
supposed to be fixed by #132 (still expecting to receive URI in the form `file:///home/mistria/git/mathjs`)
@mickaelistria could you please try with 2.0.0-pre.9?

Thanks!

**Comment by mickaelistria** [comment]
Thanks, it's working with 2.0.0-pre.9 (with `file:///` urls).
